File: DataStructure/Array/27_removeElement.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        // erase
        // auto it = nums.begin();
        // while ( it != nums.end()) {
        //     if (*it == val)
        //         nums.erase(it);
        //     else
        //         it++;
        // }
        // return nums.size();

        // 快慢指针法
        int cnt = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != val) { // 不是val 就通过慢指针覆盖 跳过val
                nums[cnt] = nums[i];
                cnt++;
            }
        }
        return cnt;
    }
};

File: DataStructure/Array/27_removeElement.java
// 双指针法-快慢指针
class Solution {
    public int removeElement(int[] nums, int val) {
        int slow = 0, fast = 0;
        while (fast < nums.length) {
            if (nums[fast] != val) {
                nums[slow] = nums[fast];
                slow++;
            }
            fast++;
        }
        return slow;
    }
}